/**
 * Copyright Soramitsu Co., Ltd. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "backend/protobuf/proto_transport_factory.hpp"
#include "backend/protobuf/proto_tx_status_factory.hpp"
#include "builders/protobuf/transaction.hpp"
#include "endpoint.pb.h"
#include "interfaces/iroha_internal/transaction_batch_factory_impl.hpp"
#include "interfaces/iroha_internal/transaction_batch_parser_impl.hpp"
#include "main/server_runner.hpp"
#include "module/irohad/ametsuchi/ametsuchi_mocks.hpp"
#include "module/irohad/multi_sig_transactions/mst_mocks.hpp"
#include "module/irohad/network/network_mocks.hpp"
#include "module/shared_model/builders/protobuf/block.hpp"
#include "module/shared_model/builders/protobuf/proposal.hpp"
#include "module/shared_model/builders/protobuf/test_block_builder.hpp"
#include "module/shared_model/builders/protobuf/test_proposal_builder.hpp"
#include "module/shared_model/builders/protobuf/test_transaction_builder.hpp"
#include "torii/command_client.hpp"
#include "torii/impl/command_service_impl.hpp"
#include "torii/impl/command_service_transport_grpc.hpp"
#include "torii/impl/status_bus_impl.hpp"
#include "torii/processor/consensus_status_processor_impl.hpp"

constexpr size_t TimesToriiBlocking = 5;

using ::testing::_;
using ::testing::A;
using ::testing::AtLeast;
using ::testing::HasSubstr;
using ::testing::Return;

using namespace iroha::network;
using namespace iroha::ametsuchi;
using namespace iroha::torii;
using namespace iroha::synchronizer;

using namespace std::chrono_literals;
constexpr std::chrono::milliseconds initial_timeout = 1s;
constexpr std::chrono::milliseconds nonfinal_timeout = 2 * 10s;

/**
The do-while cycle imitates client resubscription to the stream. Stream
"expiration" is a valid designed case (see pr #1615 for the details).

The number of attempts (3) is a magic constant here. The idea behind this number
is the following: only one resubscription is usually enough to pass the test; if
three resubscribes were not enough, then most likely there is another bug.
 */
constexpr uint32_t resubscribe_attempts = 3;

class CustomPeerCommunicationServiceMock : public PeerCommunicationService {
 public:
  CustomPeerCommunicationServiceMock(
      rxcpp::subjects::subject<
          std::shared_ptr<shared_model::interface::Proposal>> prop_notifier,
      rxcpp::subjects::subject<SynchronizationEvent> commit_notifier,
      rxcpp::subjects::subject<
          std::shared_ptr<iroha::validation::VerifiedProposalAndErrors>>
          verified_prop_notifier)
      : prop_notifier_(prop_notifier),
        commit_notifier_(commit_notifier),
        verified_prop_notifier_(verified_prop_notifier) {}

  void propagate_batch(
      std::shared_ptr<shared_model::interface::TransactionBatch> batch)
      const override {}

  rxcpp::observable<std::shared_ptr<shared_model::interface::Proposal>>
  on_proposal() const override {
    return prop_notifier_.get_observable();
  }

  rxcpp::observable<SynchronizationEvent> on_commit() const override {
    return commit_notifier_.get_observable();
  }

  rxcpp::observable<
      std::shared_ptr<iroha::validation::VerifiedProposalAndErrors>>
  on_verified_proposal() const override {
    return verified_prop_notifier_.get_observable();
  }

 private:
  rxcpp::subjects::subject<std::shared_ptr<shared_model::interface::Proposal>>
      prop_notifier_;
  rxcpp::subjects::subject<SynchronizationEvent> commit_notifier_;
  rxcpp::subjects::subject<
      std::shared_ptr<iroha::validation::VerifiedProposalAndErrors>>
      verified_prop_notifier_;
};

class ToriiServiceTest : public testing::Test {
 public:
  virtual void SetUp() {
    runner = std::make_unique<ServerRunner>(ip + ":0");

    // ----------- Command Service --------------
    pcsMock = std::make_shared<CustomPeerCommunicationServiceMock>(
        prop_notifier_, commit_notifier_, verified_prop_notifier_);
    mst = std::make_shared<iroha::MockMstProcessor>();
    wsv_query = std::make_shared<MockWsvQuery>();
    block_query = std::make_shared<MockBlockQuery>();
    storage = std::make_shared<MockStorage>();

    EXPECT_CALL(*mst, onStateUpdateImpl())
        .WillRepeatedly(Return(mst_update_notifier.get_observable()));
    EXPECT_CALL(*mst, onPreparedBatchesImpl())
        .WillRepeatedly(Return(mst_prepared_notifier.get_observable()));
    EXPECT_CALL(*mst, onExpiredBatchesImpl())
        .WillRepeatedly(Return(mst_expired_notifier.get_observable()));

    status_bus = std::make_shared<iroha::torii::StatusBusImpl>();

    EXPECT_CALL(*block_query, getTxByHashSync(_))
        .WillRepeatedly(Return(boost::none));
    EXPECT_CALL(*storage, getBlockQuery()).WillRepeatedly(Return(block_query));

    //----------- Server run ----------------
    std::unique_ptr<shared_model::validation::AbstractValidator<
        shared_model::interface::Transaction>>
        transaction_validator = std::make_unique<
            shared_model::validation::DefaultUnsignedTransactionValidator>();
    auto transaction_factory =
        std::make_shared<shared_model::proto::ProtoTransportFactory<
            shared_model::interface::Transaction,
            shared_model::proto::Transaction>>(
            std::move(transaction_validator));
    auto batch_parser =
        std::make_shared<shared_model::interface::TransactionBatchParserImpl>();
    auto batch_factory = std::make_shared<
        shared_model::interface::TransactionBatchFactoryImpl>();
    runner
        ->append(std::make_unique<torii::CommandServiceTransportGrpc>(
            std::make_shared<torii::CommandServiceImpl>(
                mst, storage, status_bus, status_factory),
            status_bus,
            initial_timeout,
            nonfinal_timeout,
            status_factory,
            transaction_factory,
            batch_parser,
            batch_factory))
        .run()
        .match(
            [this](iroha::expected::Value<int> port) {
              this->port = port.value;
            },
            [](iroha::expected::Error<std::string> err) {
              FAIL() << err.error;
            });

    runner->waitForServersReady();
  }

  std::unique_ptr<ServerRunner> runner;

  std::shared_ptr<MockWsvQuery> wsv_query;
  std::shared_ptr<MockBlockQuery> block_query;
  std::shared_ptr<MockStorage> storage;

  rxcpp::subjects::subject<std::shared_ptr<shared_model::interface::Proposal>>
      prop_notifier_;
  rxcpp::subjects::subject<SynchronizationEvent> commit_notifier_;
  rxcpp::subjects::subject<
      std::shared_ptr<iroha::validation::VerifiedProposalAndErrors>>
      verified_prop_notifier_;
  rxcpp::subjects::subject<std::shared_ptr<iroha::MstState>>
      mst_update_notifier;
  rxcpp::subjects::subject<iroha::DataType> mst_prepared_notifier;
  rxcpp::subjects::subject<iroha::DataType> mst_expired_notifier;

  std::shared_ptr<CustomPeerCommunicationServiceMock> pcsMock;
  std::shared_ptr<iroha::MockMstProcessor> mst;

  shared_model::crypto::Keypair keypair =
      shared_model::crypto::DefaultCryptoAlgorithmType::generateKeypair();

  std::shared_ptr<iroha::torii::StatusBusImpl> status_bus;
  std::shared_ptr<shared_model::proto::ProtoTxStatusFactory> status_factory =
      std::make_shared<shared_model::proto::ProtoTxStatusFactory>();

  const std::string ip = "127.0.0.1";
  int port;
};

/**
 * @given chain of CommandClient copies and moves
 * @when status is asked
 * @then grpc returns ok
 */
TEST_F(ToriiServiceTest, CommandClient) {
  iroha::protocol::TxStatusRequest tx_request;
  tx_request.set_tx_hash(std::string(32, '1'));
  iroha::protocol::ToriiResponse toriiResponse;

  auto client1 = torii::CommandSyncClient(ip, port);
  // Copy ctor
  torii::CommandSyncClient client2(client1);
  // copy assignment
  auto client3 = client2;
  // move ctor
  torii::CommandSyncClient client4(std::move(client3));
  // move assignment
  auto client5 = std::move(client4);
  auto stat = client5.Status(tx_request, toriiResponse);

  ASSERT_TRUE(stat.ok());
}
/**
 * @given torii service and number of transactions
 * @when retrieving their status
 * @then ensure those are not received
 */
TEST_F(ToriiServiceTest, StatusWhenTxWasNotReceivedBlocking) {
  std::vector<shared_model::proto::Transaction> txs;
  std::vector<shared_model::interface::types::HashType> tx_hashes;

  // create transactions, but do not send them
  for (size_t i = 0; i < TimesToriiBlocking; ++i) {
    auto tx = TestTransactionBuilder().creatorAccountId("accountA").build();
    txs.push_back(tx);
    tx_hashes.push_back(tx.hash());
  }

  // get statuses of unsent transactions
  auto client = torii::CommandSyncClient(ip, port);

  for (size_t i = 0; i < TimesToriiBlocking; ++i) {
    iroha::protocol::TxStatusRequest tx_request;
    tx_request.set_tx_hash(
        shared_model::crypto::toBinaryString(tx_hashes.at(i)));
    iroha::protocol::ToriiResponse toriiResponse;
    // this test does not require the fix for thread scheduling issues
    client.Status(tx_request, toriiResponse);
    ASSERT_EQ(toriiResponse.tx_status(),
              iroha::protocol::TxStatus::NOT_RECEIVED);
  }
}

/**
 * That test simulates the real behavior of the blocking Torii.
 *
 * @given torii service with mocked CommunicationService
 *        that is subscribed on custom rxcpp::subjects
 * @when sending some number of transactions to the Torii
 * @then ensure it perform as real one:
 *       - Torii returns ok status
 *       - Proper txs responses in Status are STATELESS_VALIDATION_SUCCESS,
 *         then STATEFUL_VALIDATION_SUCCESS and COMMITTED (order matters)
 *       - Tx that not in a block returns STATELESS_VALIDATION_SUCCESS and
           then STATEFUL_VALIDATION_FAILED
 */
TEST_F(ToriiServiceTest, StatusWhenBlocking) {
  std::vector<shared_model::proto::Transaction> txs;
  std::vector<shared_model::interface::types::HashType> tx_hashes;

  auto client1 = torii::CommandSyncClient(ip, port);

  // create transactions and send them to Torii
  std::string account_id = "some@account";
  auto now = iroha::time::now();
  for (size_t i = 0; i < TimesToriiBlocking; ++i) {
    auto shm_tx = shared_model::proto::TransactionBuilder()
                      .creatorAccountId(account_id)
                      .createdTime(now + i)
                      .setAccountQuorum(account_id, 2)
                      .quorum(1)
                      .build()
                      .signAndAddSignature(
                          shared_model::crypto::DefaultCryptoAlgorithmType::
                              generateKeypair())
                      .finish();
    const auto &new_tx = shm_tx.getTransport();

    auto stat = client1.Torii(new_tx);
    txs.push_back(shm_tx);
    tx_hashes.push_back(shm_tx.hash());

    ASSERT_TRUE(stat.ok());
  }

  torii::CommandSyncClient client2(client1);

  // emulates stateless work work
  std::for_each(tx_hashes.begin(), tx_hashes.end(), [this](auto const &hash) {
    status_bus->publish(status_factory->makeStatelessValid(
        hash, status_factory->emptyErrorMassage()));
  });

  // check if stateless validation passed
  for (size_t i = 0; i < TimesToriiBlocking; ++i) {
    iroha::protocol::TxStatusRequest tx_request;
    tx_request.set_tx_hash(
        shared_model::crypto::toBinaryString(tx_hashes.at(i)));
    iroha::protocol::ToriiResponse toriiResponse;
    client2.Status(tx_request, toriiResponse);

    ASSERT_EQ(toriiResponse.tx_status(),
              iroha::protocol::TxStatus::STATELESS_VALIDATION_SUCCESS);
  }

  // emulate the agreement on all transactions exepts last
  std::for_each(
      tx_hashes.begin(), tx_hashes.end() - 1, [this](auto const &hash) {
        status_bus->publish(status_factory->makeStatefulValid(
            hash, status_factory->emptyErrorMassage()));
      });

  auto failed_tx_hash = txs.back().hash();
  auto stringified_error = "Stateful validation error in transaction "
                         + failed_tx_hash.hex() + ": "
                                                  "command 'FailedCommand' with index '2' "
                                                  "did not pass verification with error 'stateful "
                                                  "validation failed'";
  status_bus->publish(
      status_factory->makeStatefulFail(failed_tx_hash, stringified_error));

  auto client3 = client2;
  // check if all transactions but the last one passed stateful validation
  for (size_t i = 0; i < TimesToriiBlocking - 1; ++i) {
    iroha::protocol::TxStatusRequest tx_request;
    tx_request.set_tx_hash(
        shared_model::crypto::toBinaryString(tx_hashes.at(i)));
    iroha::protocol::ToriiResponse toriiResponse;

    auto resub_counter(resubscribe_attempts);
    do {
      client3.Status(tx_request, toriiResponse);
    } while (toriiResponse.tx_status()
                 != iroha::protocol::TxStatus::STATEFUL_VALIDATION_SUCCESS
             and --resub_counter);

    ASSERT_EQ(toriiResponse.tx_status(),
              iroha::protocol::TxStatus::STATEFUL_VALIDATION_SUCCESS);
  }

  // emulate the commit status
  std::for_each(
      tx_hashes.begin(), tx_hashes.end() - 1, [this](auto const &hash) {
        status_bus->publish(status_factory->makeCommitted(
            hash, status_factory->emptyErrorMassage()));
      });

  auto client4 = client3;
  // check if all transactions but the last have committed state
  for (size_t i = 0; i < TimesToriiBlocking - 1; ++i) {
    iroha::protocol::TxStatusRequest tx_request;
    tx_request.set_tx_hash(
        shared_model::crypto::toBinaryString(tx_hashes.at(i)));
    iroha::protocol::ToriiResponse toriiResponse;
    client4.Status(tx_request, toriiResponse);

    ASSERT_EQ(toriiResponse.tx_status(), iroha::protocol::TxStatus::COMMITTED);
  }

  torii::CommandSyncClient client5(client4);
  // check if the last transaction from txs has failed stateful validation
  iroha::protocol::TxStatusRequest last_tx_request;
  last_tx_request.set_tx_hash(shared_model::crypto::toBinaryString(
      tx_hashes.at(TimesToriiBlocking - 1)));
  iroha::protocol::ToriiResponse stful_invalid_response;
  client5.Status(last_tx_request, stful_invalid_response);
  ASSERT_EQ(stful_invalid_response.tx_status(),
            iroha::protocol::TxStatus::STATEFUL_VALIDATION_FAILED);
  ASSERT_EQ(stful_invalid_response.error_message(), stringified_error);
}

/**
 * @given torii service and some number of transactions with hashes
 * @when sending request on this txs
 * @then ensure that response has the same hash as sent tx
 */
TEST_F(ToriiServiceTest, CheckHash) {
  // given
  std::vector<shared_model::interface::types::HashType> tx_hashes;
  const int tx_num = 10;

  // create transactions, but do not send them
  for (size_t i = 0; i < tx_num; ++i) {
    auto tx = TestTransactionBuilder().build();
    tx_hashes.push_back(tx.hash());
  }

  auto client = torii::CommandSyncClient(ip, port);

  // get statuses of transactions
  for (auto &hash : tx_hashes) {
    iroha::protocol::TxStatusRequest tx_request;
    tx_request.set_tx_hash(shared_model::crypto::toBinaryString(hash));
    iroha::protocol::ToriiResponse toriiResponse;
    const auto binary_hash = shared_model::crypto::toBinaryString(hash);
    auto resub_counter(resubscribe_attempts);
    do {
      client.Status(tx_request, toriiResponse);
    } while (toriiResponse.tx_hash() != binary_hash and --resub_counter);
    ASSERT_EQ(toriiResponse.tx_hash(), binary_hash);
  }
}

/**
 * @given torii service and one valid transaction
 * @when starting StatusStream and then sending transaction to Iroha
 * @then ensure that response will have at least 3 statuses
 * (it should contain STATELESS_VALIDATION_SUCCESS, STATEFUL_VALIDATION_SUCCESS
 * and COMMITTED) and the last status should be COMMITTED
 */
TEST_F(ToriiServiceTest, StreamingFullPipelineTest) {
  using namespace shared_model;

  auto client = torii::CommandSyncClient(ip, port);
  auto iroha_tx = proto::TransactionBuilder()
                      .creatorAccountId("a@domain")
                      .setAccountQuorum("a@domain", 2)
                      .createdTime(iroha::time::now())
                      .quorum(1)
                      .build()
                      .signAndAddSignature(keypair)
                      .finish();

  std::string txhash = crypto::toBinaryString(iroha_tx.hash());

  std::vector<iroha::protocol::ToriiResponse> torii_response;
  // StatusStream is a blocking call and returns only when the last status
  // (Committed in this case) will be received. We start request before
  // transaction sending so we need in a separate thread for it.
  std::thread t([&] {
    auto resub_counter(resubscribe_attempts);
    iroha::protocol::TxStatusRequest tx_request;
    tx_request.set_tx_hash(txhash);
    do {
      client.StatusStream(tx_request, torii_response);
    } while (torii_response.back().tx_status()
                 != iroha::protocol::TxStatus::COMMITTED
             and --resub_counter);
  });

  client.Torii(iroha_tx.getTransport());
  status_bus->publish(status_factory->makeStatelessValid(
      iroha_tx.hash(), status_factory->emptyErrorMassage()));
  status_bus->publish(status_factory->makeStatefulValid(
      iroha_tx.hash(), status_factory->emptyErrorMassage()));
  status_bus->publish(status_factory->makeCommitted(
      iroha_tx.hash(), status_factory->emptyErrorMassage()));
  t.join();

  // we can be sure only about final status
  // it can be only one or to follow by some non-final
  ASSERT_EQ(torii_response.back().tx_status(),
            iroha::protocol::TxStatus::COMMITTED);
}

/**
 * @given torii service and fake transaction hash
 * @when sending streaming request for this hash
 * @then ensure that response will have exactly 1 status - NOT_RECEIVED
 */
TEST_F(ToriiServiceTest, StreamingNoTx) {
  auto client = torii::CommandSyncClient(ip, port);
  std::vector<iroha::protocol::ToriiResponse> torii_response;
  std::thread t([&]() {
    iroha::protocol::TxStatusRequest tx_request;
    tx_request.set_tx_hash("0123456789abcdef");
    // this test does not require the fix for thread scheduling issues
    client.StatusStream(tx_request, torii_response);
  });

  t.join();
  ASSERT_EQ(torii_response.size(), 1);
  ASSERT_EQ(torii_response.at(0).tx_status(),
            iroha::protocol::TxStatus::NOT_RECEIVED);
}

/**
 * Checks that torii is able to handle lists (sequences) of transactions
 *
 * @given torii service and collection of transactions
 * @when that collection is asked to be processed by Torii
 * @then statuses of all transactions from that request are
 * ENOUGH_SIGNATURES_COLLECTED
 */
TEST_F(ToriiServiceTest, ListOfTxs) {
  const auto test_txs_number = 5;

  // initial preparations: creation of variables and txs
  auto client = torii::CommandSyncClient(ip, port);
  std::vector<shared_model::interface::types::HashType> tx_hashes(
      test_txs_number);
  iroha::protocol::TxList tx_list;

  for (auto i = 0; i < test_txs_number; ++i) {
    auto shm_tx = shared_model::proto::TransactionBuilder()
                      .creatorAccountId("doge@master" + std::to_string(i))
                      .createdTime(iroha::time::now())
                      .setAccountQuorum("doge@master", 2)
                      .quorum(1)
                      .build()
                      .signAndAddSignature(
                          shared_model::crypto::DefaultCryptoAlgorithmType::
                              generateKeypair())
                      .finish();
    tx_hashes[i] = shm_tx.hash();
    new (tx_list.add_transactions())
        iroha::protocol::Transaction(shm_tx.getTransport());
  }

  // send the txs
  client.ListTorii(tx_list);

  // emulate work
  std::for_each(tx_hashes.begin(), tx_hashes.end(), [this](const auto &hash) {
    status_bus->publish(status_factory->makeEnoughSignaturesCollected(
        hash, status_factory->emptyErrorMassage()));
  });

  // check their statuses
  std::for_each(
      std::begin(tx_hashes), std::end(tx_hashes), [&client](auto &hash) {
        iroha::protocol::TxStatusRequest tx_request;
        tx_request.set_tx_hash(shared_model::crypto::toBinaryString(hash));
        iroha::protocol::ToriiResponse toriiResponse;

        auto resub_counter(resubscribe_attempts);
        do {
          client.Status(tx_request, toriiResponse);
        } while (toriiResponse.tx_status()
                     != iroha::protocol::TxStatus::ENOUGH_SIGNATURES_COLLECTED
                 and --resub_counter);

        ASSERT_EQ(toriiResponse.tx_status(),
                  iroha::protocol::TxStatus::ENOUGH_SIGNATURES_COLLECTED);
      });
}

/**
 * Checks that in case of failed transactions list every transaction has a
 * related status and error message
 *
 * @given torii service and a list of bad-formed transactions
 * @when checking those transactions after sending them to Torii
 * @then every transaction from this list will have STATELESS_FAILED status @and
 * the same corresponding error message
 */
TEST_F(ToriiServiceTest, FailedListOfTxs) {
  const auto test_txs_number = 5;

  // initial preparations: creation of variables and txs
  auto client = torii::CommandSyncClient(ip, port);
  std::vector<shared_model::interface::types::HashType> tx_hashes(
      test_txs_number);
  iroha::protocol::TxList tx_list;

  for (auto i = 0; i < test_txs_number; ++i) {
    auto shm_tx = TestTransactionBuilder()
                      .creatorAccountId("doge@master" + std::to_string(i))
                      .createdTime(iroha::time::now(std::chrono::hours(24)
                                                    + std::chrono::minutes(1)))
                      .setAccountQuorum("doge@master", 2)
                      .quorum(1)
                      .build();
    tx_hashes[i] = shm_tx.hash();
    new (tx_list.add_transactions())
        iroha::protocol::Transaction(shm_tx.getTransport());
  }

  // send the txs
  client.ListTorii(tx_list);

  // check their statuses
  std::for_each(
      std::begin(tx_hashes), std::end(tx_hashes), [&client](auto &hash) {
        iroha::protocol::TxStatusRequest tx_request;
        tx_request.set_tx_hash(shared_model::crypto::toBinaryString(hash));
        iroha::protocol::ToriiResponse toriiResponse;
        auto resub_counter(resubscribe_attempts);
        do {
          client.Status(tx_request, toriiResponse);
        } while (toriiResponse.tx_status()
                     != iroha::protocol::TxStatus::STATELESS_VALIDATION_FAILED
                 and --resub_counter);

        ASSERT_EQ(toriiResponse.tx_status(),
                  iroha::protocol::TxStatus::STATELESS_VALIDATION_FAILED);
        auto msg = toriiResponse.error_message();
        ASSERT_THAT(msg, HasSubstr("bad timestamp: sent from future"));
      });
}
